lib/internal/util/safe-property-assignment.js

```diff
@@ -0,0 +1,64 @@
+'use strict';
+
+const {
+  ObjectCreate,
+  ObjectDefineProperties,
+  ObjectPrototypeHasOwnProperty,
+  ReflectDefineProperty,
+  ReflectOwnKeys,
+} = primordials;
+
+/**
+ * Mimics `obj[key] = value` but ignoring potential prototype inheritance.
+ * @param {any} obj
+ * @param {string} key
+ * @param {any} value
+ * @returns {boolean}
+ */
+function setOwnProperty(obj, key, value) {
+  if (ObjectPrototypeHasOwnProperty(obj, key)) {
+    obj[key] = value;
+    return true;
+  }
+  return ReflectDefineProperty(obj, key, {
+    __proto__: null,
+    configurable: true,
+    enumerable: true,
+    value,
+    writable: true,
+  });
+}
+
+/**
+ * Mimics `Object.assign` but ignoring potential prototype inheritance.
+ * @template T
+ * @param {T} obj
+ * @param {Record<string|symbol, unknown>[]} sources
+ * @returns {T}
+ */
+function assignOwnProperties(obj, ...sources) {
+  const descriptors = ObjectCreate(null);
+  for (let i = 0; i < sources.length; i++) {
+    const keys = ReflectOwnKeys(sources[i]);
+    for (let j = 0; j < keys.length; j++) {
+      const key = keys[j];
+      if (ObjectPrototypeHasOwnProperty(obj, key)) {
+        obj[key] = sources[i][key];
+      } else {
+        descriptors[key] = {
+          __proto__: null,
+          configurable: true,
+          enumerable: true,
+          value: sources[i][key],
+          writable: true,
+        };
+      }
```

Comment:
won't this change the observable key ordering, by assigning the own properties immediately but defining the non-own properties later?

Reply:
If it's already defined, it wouldn't affect the original ordering, would it?

```js
console.log(JSON.stringify(assignOwnProperties({ a: 23, b: 27 }, { c: 42, test: 1, d: 89 }))); 
// {"a":23,"b":27,"c":42,"test":1,"d":89}

console.log(JSON.stringify(assignOwnProperties({ a: 23, test: 0, b: 27 }, { c: 42, test: 1, d: 89 })));
// {"a":23,"test":1,"b":27,"c":42,"d":89}
```

Reply:
I mean, compared to the previous approach - like wouldn't this make places that refactor to use this helper suddenly have a different key ordering?

Reply:
```js
console.log(JSON.stringify(assignOwnProperties({ a: 23, b: 27 }, { c: 42, test: 1, d: 89 })));
// {"a":23,"b":27,"c":42,"test":1,"d":89}
console.log(JSON.stringify(Object.assign({ a: 23, b: 27 }, { c: 42, test: 1, d: 89 })));
// {"a":23,"b":27,"c":42,"test":1,"d":89}
{
    const obj = { a: 23, b: 27 }
    obj.c = 42
    obj.test = 1
    obj.d = 89
    
    console.log(JSON.stringify(obj)); // {"a":23,"b":27,"c":42,"test":1,"d":89}
}
console.log(JSON.stringify(assignOwnProperties({ a: 23, test: 0, b: 27 }, { c: 42, test: 1, d: 89 })));
// {"a":23,"test":1,"b":27,"c":42,"d":89}
console.log(JSON.stringify(Object.assign({ a: 23, test: 0, b: 27 }, { c: 42, test: 1, d: 89 })));
// {"a":23,"test":1,"b":27,"c":42,"d":89}
{
    const obj = { a: 23, test: 0, b: 27 }
    obj.c = 42
    obj.test = 1
    obj.d = 89
    
    console.log(JSON.stringify(obj)); // {"a":23,"test":1,"b":27,"c":42,"d":89}
}
```

Reply:
Thanks for the confirmation :-)

is the ordering the same when you refactored normal assignments? like the `code` property on Errors, eg

Reply:
Not always, but that should be fixed by https://github.com/nodejs/node/pull/46065/commits/2e892fcfe9569b2be2a429cf1390ca6d432e2cd3 :)